### PR TITLE
More user-friendly server rejection messages

### PR DESCRIPTION
**Pull request by Wuzzy2**
I thought some of the messages which the server sends when it rejects a clients could be more useful. 
Frankly, the current messages don't convey much information and were a bit poorly worded. So I attempted to fix that. :-)

### CODE REVIEW

**Comment by PilzAdam** [comment]
Your messages are just more verbose, except for the allowed characters.

**Comment by est31** [comment]
Yes, this is mostly a preference question. I like it the current way. Listing the allowed characters is an useful addition though.

File: src/network/networkprotocol.h [hunk 1]
```diff
@@ -942,16 +942,16 @@ enum NetProtoCompressionMode {
...
-	"Empty passwords are disallowed.  Set a password and try again.",
-	"Another client is connected with this name.  If your client closed unexpectedly, try again in a minute.",
-	"Server authention failed.  This is likely a server error."
+	"You have entered a wrong password.",
+	"Your client sent something the server didn't expect. Try reconnecting or updating your client.",
+	"The server is running in simple singleplayer mode. You cannot connect.",
+	"Your client's version is not supported.\nPlease contact a server administrator.",
+	"The player name contains unallowed characters. Allowed characters are a-z, A-Z, 0-9, hyphen and underscore.",
+	"The player name is not allowed.",
+	"There are currently too many users on this server and it won't accept further users. Please wait until a few users have left.",
```
Comment (rubenwardy) [inline src/network/networkprotocol.h]: This should be made more consice, you can remove " and it won't accept further users".

It also is longer than 80 chars.

**Comment by rubenwardy** [comment]
I agree with some of these, but some of your new messages are too verbose.

**Comment by kilbith** [comment]
Agree with the aforementioned points. Exactly the same problem that had : https://github.com/minetest/minetest/pull/2438 (i.e. you tend to write too much).

**Comment by Wuzzy2** [comment]
The new messages are “not just more verbose”. The previous sentences were IMO poorly worded and had no complete sentences, etc.

Remember, those messages are directed to the end user, not to developers who already know what lies “behind” a certain message.

I will rebase everything as soon you are happy with the current stage.

**Comment by C1ffisme** [comment]
Too verbose? Where did you all grow up, text talk school? What would you rather prefer seing after you have typed your 50-character password wrong?

A. "Invalid password"
B. "Sorry, but you may have gotten your password wrong. Try seeing if Caps Lock or Numb Lock is on and try again."
C. "Try Agn."

I imagine you would want to see number B, and not letter A or symbol C. If you think C, then what I have said is true. You did live in text talk school.

**Comment by est31** [comment]
shorter names are better when you are talking about them.
<quote>Hi, I'm trying to log in to your server, but I'm getting "Sorry, but you may have gotten your password wrong. Try seeing if Caps Lock or Numb Lock is on and try again.". Can you help me?
</quote>

**Comment by est31** [comment]
IMO "telegram style" is good.

A suggestion to resolve this discussion, where most core devs are on the "telegram style" side:
we make them localizable (thats why they are an array the first place), and then @Wuzzy2 can add a "en-verbose" locale. Everybody who likes spending their time reading error messages can switch to that locale then.

**Comment by C1ffisme** [comment]
True, but people should just learn to actually type. There are people who would type text talk on COMPUTERS because they just don't have the wish to type. Want proof? Think of all the suggestions for things on forums that get a "No Support" rather than "I'm sorry, but (reason). Maybe (helpful criticism)".

For people who don't want to type? Add an "Error (number):" before each error. That way you can say "Hey, I got an Error 51, can you help?"

**Comment by rubenwardy** [comment]
@C1ffisme none of the options. I'd go with:

> Wrong password. Check your caps lock and try again.

**Comment by C1ffisme** [comment]
@rubenwardy Still sounds like the computer is being rude to you. :P

File: src/network/serverpackethandler.cpp [hunk 2]
```diff
@@ -526,7 +530,9 @@ void Server::handleCommand_Init_Legacy(NetworkPacket* pkt)
 		actionstream << "Server: " << playername << " tried to join, but there"
 				<< " are already max_users="
 				<< g_settings->getU16("max_users") << " players." << std::endl;
-		DenyAccess_Legacy(pkt->getPeerId(), L"Too many users.");
+		DenyAccess_Legacy(pkt->getPeerId(), L"There are currently too many users on this"
```
Comment (rubenwardy) [inline src/network/serverpackethandler.cpp]: This is different from https://github.com/minetest/minetest/pull/2786/files#diff-d6c71fb6c2c5e2ad979f39a0422f7786R951

**Comment by rubenwardy** [comment]
> Think of all the suggestions for things on forums that get a "No Support"

What do you even mean?

------------
> Still sounds like the computer is being rude to you. :P 

I think it sounds matter of fact.
------------

There does need to be compromise between being concise and easy to understand.

File: src/network/networkprotocol.h [hunk 3]
```diff
@@ -942,16 +942,16 @@ enum NetProtoCompressionMode {
...
-	"Player name not allowed.",
-	"Too many users.",
-	"Empty passwords are disallowed.  Set a password and try again.",
-	"Another client is connected with this name.  If your client closed unexpectedly, try again in a minute.",
-	"Server authention failed.  This is likely a server error."
+	"You have entered a wrong password.",
+	"Your client sent something the server didn't expect. Try reconnecting or updating your client.",
+	"The server is running in simple singleplayer mode. You cannot connect.",
+	"Your client's version is not supported.\nPlease contact a server administrator.",
+	"The player name contains unallowed characters. Allowed characters are a-z, A-Z, 0-9, hyphen and underscore.",
```
Comment (rubenwardy) [inline src/network/networkprotocol.h]: unallowed isn't a word: http://dictionary.cambridge.org/spellcheck/american-english/?q=unallowed
disallowed is correct.

File: src/network/serverpackethandler.cpp [hunk 4]
```diff
@@ -462,21 +463,24 @@ void Server::handleCommand_Init_Legacy(NetworkPacket* pkt)
...
-		DenyAccess_Legacy(pkt->getPeerId(), L"Empty name");
+		DenyAccess_Legacy(pkt->getPeerId(), L"The name is empty.");
 		return;
 	}
 
 	if (string_allowed(playername, PLAYERNAME_ALLOWED_CHARS) == false) {
 		actionstream << "Server: Player with an invalid name "
 				<< "tried to connect from " << addr_s << std::endl;
-		DenyAccess_Legacy(pkt->getPeerId(), L"Name contains unallowed characters");
+		DenyAccess_Legacy(pkt->getPeerId(), L"The chosen name contains unallowed characters."
```
Comment (rubenwardy) [inline src/network/serverpackethandler.cpp]: same here

**Comment by nerzhul** [comment]
Be precise and concise. It's error messages, the goal is to understand the problem, no talk about minetest life :).

Some messages needs to be rewritten better, like the unallowed problem mentionned by @rubenwardy , i agree

**Comment by kwolekr** [comment]
The current messages get the point across just fine.  Apart from the name length limit and typo fix, what this PR does isn't needed IMHO.  More verbose != more understandable.

Thumbs down from me.  Sorry. :-1:

File: src/network/networkprotocol.h [hunk 5]
```diff
@@ -942,16 +942,16 @@ enum NetProtoCompressionMode {
...
-	"Your client sent something the server didn't expect.  Try reconnecting or updating your client",
-	"The server is running in simple singleplayer mode.  You cannot connect.",
-	"Your client's version is not supported.\nPlease contact server administrator.",
-	"Player name contains disallowed characters.",
-	"Player name not allowed.",
-	"Too many users.",
-	"Empty passwords are disallowed.  Set a password and try again.",
-	"Another client is connected with this name.  If your client closed unexpectedly, try again in a minute.",
-	"Server authention failed.  This is likely a server error."
+	"You have entered a wrong password.",
```
Comment (ShadowNinja) [inline src/network/networkprotocol.h]: I don't think `wrong` is gramatically correct here.  "The password you entered is wrong." would be better, but I'd just use `incorrect` or `invalid`.

File: src/network/networkprotocol.h [hunk 6]
```diff
@@ -942,16 +942,16 @@ enum NetProtoCompressionMode {
...
+	"Your client sent something the server didn't expect. Try reconnecting or updating your client.",
+	"The server is running in simple singleplayer mode. You cannot connect.",
+	"Your client's version is not supported.\nPlease contact a server administrator.",
+	"The player name contains disallowed characters. Allowed characters are a-z, A-Z, 0-9, hyphen and underscore.",
```
Comment (ShadowNinja) [inline src/network/networkprotocol.h]: `The` -> `Your`
Also, add a comment next to `PLAYERNAME_ALLOWED_CHARS` noting that this has to be updated it that's changed.  I hink there's a `PLAYERNAME_ALLOWED_CHARS_HUMAN_READABLE` or similar macro that you should use instead though.

File: src/network/networkprotocol.h [hunk 7]
```diff
@@ -942,16 +942,16 @@ enum NetProtoCompressionMode {
...
+	"The player name is not allowed.",
```
Comment (ShadowNinja) [inline src/network/networkprotocol.h]: `The` -> `Your`

File: src/network/serverpackethandler.cpp [hunk 8]
```diff
@@ -454,7 +454,8 @@ void Server::handleCommand_Init_Legacy(NetworkPacket* pkt)
 	if (playername_length == PLAYERNAME_SIZE) {
 		actionstream << "Server: Player with name exceeding max length "
 				<< "tried to connect from " << addr_s << std::endl;
-		DenyAccess_Legacy(pkt->getPeerId(), L"Name too long");
+		DenyAccess_Legacy(pkt->getPeerId(), L"This name is too long. Maximum allowed length: "
```
Comment (ShadowNinja) [inline src/network/serverpackethandler.cpp]: `This` -> `Your`

File: src/network/serverpackethandler.cpp [hunk 9]
```diff
@@ -462,21 +463,24 @@ void Server::handleCommand_Init_Legacy(NetworkPacket* pkt)
 	if (playername[0]=='\0') {
 		actionstream << "Server: Player with an empty name "
 				<< "tried to connect from " << addr_s << std::endl;
-		DenyAccess_Legacy(pkt->getPeerId(), L"Empty name");
+		DenyAccess_Legacy(pkt->getPeerId(), L"The name is empty.");
```
Comment (ShadowNinja) [inline src/network/serverpackethandler.cpp]: `The` -> `Your`

File: src/network/serverpackethandler.cpp [hunk 10]
```diff
@@ -462,21 +463,24 @@ void Server::handleCommand_Init_Legacy(NetworkPacket* pkt)
...
 		return;
 	}
 
 	if (string_allowed(playername, PLAYERNAME_ALLOWED_CHARS) == false) {
 		actionstream << "Server: Player with an invalid name "
 				<< "tried to connect from " << addr_s << std::endl;
-		DenyAccess_Legacy(pkt->getPeerId(), L"Name contains unallowed characters");
+		DenyAccess_Legacy(pkt->getPeerId(), L"The chosen name contains disallowed characters."
+				L" Allowed characters are a-z, A-Z, 0-9, hyphen and underscore.");
```
Comment (ShadowNinja) [inline src/network/serverpackethandler.cpp]: As before, also `The` -> `Your`.

File: src/network/serverpackethandler.cpp [hunk 11]
```diff
@@ -462,21 +463,24 @@ void Server::handleCommand_Init_Legacy(NetworkPacket* pkt)
...
 		return;
 	}
 
 	if (!isSingleplayer() && strcasecmp(playername, "singleplayer") == 0) {
 		actionstream << "Server: Player with the name \"singleplayer\" "
 				<< "tried to connect from " << addr_s << std::endl;
-		DenyAccess_Legacy(pkt->getPeerId(), L"Name is not allowed");
+		DenyAccess_Legacy(pkt->getPeerId(), L"The name \"singleplayer\" is a special name"
+				L" which is reserved for singleplayer games. It can"
+				L" not be used on multiplayer servers.");
```
Comment (ShadowNinja) [inline src/network/serverpackethandler.cpp]: This should be the same as the error message for new clients.  Consider using `accessDeniedStrings[AD_SINGLEPLAYER]` or similar.

**Comment by ShadowNinja** [comment]
Why did you remove the sentence double spacing?
Also, the logacy and new access denied messages are different, that should be fixed.

**Comment by Wuzzy2** [comment]
Sorry, I am no longer working on this. Especially if I have to justify every single space (or non-space).

**Comment by kwolekr** [comment]
Due to the general lack of interest in this PR and its abandonment by the author, I am closing it.
I think some valid points did come up from this discussion though, namely:
1). You should be able to localize these messages, but the work required for this is somewhat more involved. 
2). The ideal message format has to do with display more than text content.  The short 'telegraph style' content would be front and center, in larger text, and the finer detailed verbose explanation would be below in smaller print.  That's also outside of the scope of this PR.

In either case, if interest is renewed in this topic, I suggest creating a new issue for one (or both) of the above points.

**Pull request closed without merging**